pylint/checkers/modified_iterating_checker.py

```diff
@@ -0,0 +1,134 @@
+# Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+# For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
+
+from typing import TYPE_CHECKING
+
+import astroid
+from astroid import nodes
+
+from pylint import checkers, interfaces
+from pylint.checkers import utils
+
+if TYPE_CHECKING:
+    from pylint.lint import PyLinter
+
+
+class ModifiedIterationChecker(checkers.BaseChecker):
+    """Checks for modified for loops iterations
+    Currently supports `for` loops for Sets, Dictionaries and Lists.
+    """
+
+    __implements__ = interfaces.IAstroidChecker
+
+    name = "modified_iteration"
+
+    msgs = {
+        "W4701": (
+            "Iterated list '%s' is being modified inside loop body, consider iterating through a copy of it instead.",
+            "modified-iterating-list",
+            "Emitted when items are added or removed to a list being iterated through. "
+            "Doing so ccan result in unexpected behaviour, that is why it is preferred to use a copy of the list.",
+        ),
+        "E4702": (
+            "Iterated dict '%s' is being modified inside loop body, iterate through a copy of it instead.",
+            "modified-iterating-dict",
+            "Emitted when items are added or removed to a dict being iterated through. "
+            "Doing so raises a RuntimeError",
+        ),
+        "E4703": (
+            "Iterated set '%s' is being modified inside loop body, iterate through a copy of it instead.",
+            "modified-iterating-set",
+            "Emitted when items are added or removed to a set being iterated through. "
+            "Doing so raises a RuntimeError",
+        ),
+    }
+
+    options = ()
+    priority = -2
+
+    @utils.check_messages(
+        "modified-iterating-list", "modified-iterating-dict", "modified-iterating-set"
+    )
+    def visit_for(self, node: nodes.For) -> None:
+        iter_obj = node.iter
+        for body_node in node.body:
+            msg_id = None
+            if self._modified_iterating_list_cond(body_node, iter_obj):
+                msg_id = "modified-iterating-list"
+            elif self._modified_iterating_dict_cond(body_node, iter_obj):
+                msg_id = "modified-iterating-dict"
+            elif self._modified_iterating_set_cond(body_node, iter_obj):
+                msg_id = "modified-iterating-set"
+            if msg_id is not None:
+                self.add_message(
+                    msg_id,
+                    node=node,
+                    args=(iter_obj.name,),
+                )
+                break  # since the msg is raised for the `for` node no further check is needed
```

Comment:
I don't particularly like this: If `modified-iterating-list` is being ignored in `pylintrc` and `modified-iterating-set` should also be raised it won't raise the message any longer and there will be no messages at all.

Personally I always prefer raising every correct message as you never know how users disable their code. If both a `dict` and a `set` are being modified we should warn accordingly.
I wonder if this is needed though: how can you iterate over a `set` and a `dict` at the same time?

Reply:
@DanielNoord  - I understand what you mean but I think you are right, every `for` loop is considered separately and every `for` loop can contain only 1 list/dict/set.
and for the rare case that we iterate a `list` and a `set` or `dict` in the SAME `for` loop, if it exist, it wouldn't be detected anyway probably (because the structure of such `for` loop is not supported in my code).

Reply:
Do we need the `break` statement here then? If we don't consider the case where we iterate over two types of iterators this `break` doesn't do much for us right?